### ISSUE TITLE
Be more explicit about how we use memory

### ISSUE THREAD

**Issue by SamRoberts** [SamRoberts/fuzzy_rust#9]
The fuzzy algorithm was initially written fairly quickly with only a passing concern for what our memory use looked like. As such:

- there is no consistent principle governing the use of copied values vs. references to values other then "hey, this struct looks pretty small and immutable", and
- there might be a nice way to avoid creating a new `Vec<Next>` every time we want to consider the possible future steps from a given position.

Review memory use and use of references, _but only after implementing benchmarks_.

**Comment by SamRoberts** [comment]
My short-term goal for fuzzy is to complete #12. Before I do that, I want to take my own stab at being less naive about memory use / moving / copying.

**Comment by SamRoberts** [comment]
I've removed `succ` and it's temporary vector. Other then that, I think using references to patt and text items (which I did earlier) was the only thing that jumps out at me.

I could potentially pre-compute all `Ix` values (probably in the `Config`) and just pass around references to them, but it's really not obvious that this would be an improvement ... let's not do any more on this before implementing benchmarks or asking for feedback.
